Generalize `compute_MVBS` and improve its efficiency for Dask input

b-reyes opened the pull request:
This PR addresses #662 and #845 by directly calculating `MVBS` in an efficient manner. This new approach allows us to apply `compute_MVBS` on `echo_range` data sizes that vary with `ping_time` and `Sv` values that are Dask arrays. In addition to the new core routines that compute `MVBS`, unit tests for the calculation of `MVBS` were also completed. These tests construct mock `Sv` values with known `MVBS` values.  

Please note that this PR is currently a draft. The following items need to be completed: 
- [x] Finish documenting the function `get_MVBS_along_channels` in `preprocess/api.py`
- [x] Replace the previous `ds_MVBS` calculation in `compute_MVBS` with the new and improved `get_MVBS_along_channels` function 
- [x] Remove the restriction that `echo_range` must be the same for each `ping_time`
- [x] Make sure that the new `compute_MVBS` function works well with the 19 files in the OOI eclipse notebook
- [ ] ~~Possibly replace `test_compute_MVBS` with a more general test using the ideas in `test_bin_and_mean_2d`~~

b-reyes commented:
@leewujung I have went through and documented/cleaned up the code I previously mentioned. This code is now in a state that I am happy with, please go ahead and review it.

leewujung commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+
+    Parameters
+    ----------
+    arr: np.ndarray or dask.array.Array
+        2D array to bin and mean
```
```suggestion
    arr: np.ndarray or dask.array.Array
        2D array (dimension: [``echo_range`` x ``ping_time``]) to bin  along ``echo_range``
        and compute mean of each bin
```

leewujung commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    ----------
+    echo_range: np.ndarray
+        2D array of echo range values
+    bins_er: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``echo_range``
```
This should also be a 2D array since `echo_range` is 2D?
```suggestion
    bins_er: np.ndarray
        2D array (used by np.digitize) representing the binning required for ``echo_range``
        ``bins_er`` should be of the same shape as ``echo_range``
```

b-reyes replied:
As we discussed, `bin_er` should not be a 2D array. I will resolve this conversation.

leewujung commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    This function assumes that ``arr`` has rows corresponding to
+    ``ping_time`` and columns corresponding to ``echo_range``.
+
+    This function should not be run if the number of ``echo_range`` values
+    vary amongst ``ping_times``.
```
By "the number of `echo_range` values" do you mean the number of bins along echo range? In the case of data created by echopype, the scenario of the number of `echo_range` values varying across `ping_times` would never occur since we NaN-padded the shorter pings, right?

b-reyes replied:
> By "the number of echo_range values" do you mean the number of bins along echo range?

By the "the number of echo_range values" I meant that the 2D array `echo_range` should have columns that are of the same length (i.e. no ragged arrays). 

> In the case of data created by echopype, the scenario of the number of echo_range values varying across ping_times would never occur since we NaN-padded the shorter pings, right?

Yes, this should never occur for `compute_MVBS` because we padded the shorter pings. I put this comment there just incase in the future we wanted to use this function for a different application.

leewujung replied:
Ok I see. Could you please also add this second part (that this should not occur for our current use of echopype-generated Sv data) so that it is clear what this is intended for? Thanks!

b-reyes replied:
No problem, I will go ahead and add a comment here.

leewujung commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    # obtain the bin indices for echo_range and times
+    digitized_echo_range, bin_time_ind = get_bin_indices(echo_range, bins_er, times, bins_time)
+
+    # determine if grouping of echo_range values with the same step size is necessary
+    if comp_er_check:
+        grouping_needed = is_grouping_needed_comprehensive(echo_range)
+    else:
+        grouping_needed = is_grouping_needed_less_comprehensive(echo_range)
```
I have the following suggestions on refactoring/renaming the variables and functions in this block:
- rename `grouping_needed` to `er_grouping_need` just to make it more specific
- rename `comp_er_check` to `comprehensive_er_check` just to make it more specific
- rename the current 2 functions:
   - `is_grouping_needed_comprehensive` --> `if_all_er_steps_identical`
   - `is_grouping_needed_less_comprehensive` --> `if_last_er_steps_identical`
- combine the above 2 checking functions into one `is_er_grouping_needed` with `comp_er_check` as the input argument
   ```python
   def is_er_grouping_needed(echo_range, comprehensive_er_check):
      if comp_er_check:
         return if_all_er_steps_identical(echo_range)
      else:
         return if_last_er_steps_identical(echo_range)
   ```
My thought is that this way when reading this function, one can abstract away whatever the operations under the 2 checking functions are and just know that one is more comprehensive than the other, unless they are interested in knowing the details.

b-reyes replied:
All of these changes sound good to me! I will go ahead and implement them.

leewujung commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+
+    if grouping_needed:
+        # groups, bins, and means arr with respect to echo_range
+        er_means = group_bin_mean_echo_range(arr, digitized_echo_range, n_bin_er)
```
For the sake of clarify, maybe change this function name to `er_group_bin_mean_echo_range`

b-reyes replied:
Why did you add `er`? The way I read `group_bin_mean_echo_range` is: group first, bin these values, and then take the mean along `echo_range`.  If you are trying to specify that the grouping is with respect to `digitized_echo_range`, then I suggest the following name `group_dig_er_bin_mean_echo_range`.

leewujung replied:
I like `group_dig_er_bin_mean_echo_range` (even though one can say it's long :P) -- wanted to make this more specific just because it is not clear what is being grouped, since binning is already a type of grouping.

b-reyes replied:
Yes, I agree that it is a bit long, but it describes very specifically what is happening. I will go ahead and name it `group_dig_er_bin_mean_echo_range`, since you seem to be fine with it.

leewujung commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+        2D array representing the bin and mean of ``arr`` along ``echo_range``
+    """
+
+    binned_means = []
+    for bin_er in range(1, n_bin_er):
+
+        # bin and mean echo_range dimension
+        er_selected_data = np.nanmean(arr[:, digitized_echo_range == bin_er], axis=1)
+
+        # collect all echo_range bins
+        binned_means.append(er_selected_data)
```
If we wrap this line
```
er_selected_data = np.nanmean(arr[:, digitized_echo_range == bin_er], axis=1)
```
into a function and delay it, wouldn't dask be able to distributing the nanmean computation instead of computing each binned averages in a for loop?

b-reyes replied:
If the input to `compute_MVBS` is a lazy Dataset, then in this code section `arr` will also be delayed. For this reason, there is no computation being done in the for loop you highlighted and the output of this function will also be delayed. When we actually compute this, Dask will efficiently distribute the computation. If we wrapped the nanmean in a delayed function then this would lead to inefficient computation and overwhelm the scheduler with too many tasks.

leewujung replied:
> there is no computation being done in the for loop you highlighted and the output of this function will also be delayed

Yes I understood this, my question was whether this was computed sequentially or simultaneously (distributed).

Since this is a for loop, seems that Dask would distribute within each iteration of the loop, and not running all the different bins in parallel. When you say
```
When we actually compute this, Dask will efficiently distribute the computation.
```
what is being distributed efficiently?

> If we wrapped the nanmean in a delayed function then this would lead to inefficient computation and overwhelm the scheduler with too many tasks.

Do you mean that the number of tasks is the cause of inefficiency? Or the number of small tasks, whose size is related to the size of the bins along echo_range and ping_time?

leewujung commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    if isinstance(er_means, dask.array.Array):
+        er_means = er_means.compute()
+
+    # create final reduced array i.e. MVBS
+    final = np.empty((n_bin_time, n_bin_er - 1))
+    for bin_time in range(1, n_bin_time + 1):
+
+        # obtain er_mean indices corresponding to the time bin
+        indices = np.argwhere(bin_time_ind == bin_time).flatten()
+
+        if len(indices) == 0:
+            # fill values with NaN, if there are no values in the bin
+            final[bin_time - 1, :] = np.nan
+        else:
+            # bin and mean the er_mean time bin
+            final[bin_time - 1, :] = np.nanmean(er_means[:, indices], axis=1)
```
I have a similar question as in under the function `bin_and_mean_echo_range` above. Wouldn't you be able to wrap the core ops into a function and delay it, so that the binned averages are distributed instead of computed in a loop?

leewujung replied:
Or, `bin_and_mean_2d` can be delayed so that the MVBS in each channel is calculated in parallel?

leewujung reviewed (commented):
@b-reyes : I am done with my first round of review for now. I recommended some refactoring that should be trivial to do. I have some questions on why for loops instead of delayed functions are used in a couple of places when binned averages are computed. Is there a performance concern here if the bins are too small (i.e., contain too little number of values?). If that's case, I am also wondering about delaying `bin_and_mean_2d`  for each channel instead.

I have not reviewed the tests and also the content of the 2 `is_grouping_needed_*` functions, but I can do that in the next round. Wanted to return this to you first so that I don't hold up the process. Thanks.

leewujung commented:
Ok @b-reyes and I resolved the main comments above and have the 2 items:
- @leewujung : to add an issue for later enhancement re using either the temp zarr store approach or the scatter approach for the intermediate step from `er_means` to ping time bins
- @b-reyes : to move most of the MVBS functions to a separate script (i.e. out of api.py)

b-reyes commented on echopype/preprocess/api.py:
```diff
@@ -329,3 +286,433 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    for bin_er in range(1, n_bin_er):
+
+        # Catch a known warning that can occur, which does not impact the results
+        with warnings.catch_warnings():
```
@leewujung When we were running the 19 OOI files on data that was in memory, I noticed that a warning was caused by this code block. This warning was spit out by `nanmean` because it was taking the mean of an array filled with NaNs. This still produces the correct output, but the warning is produced. For this reason, I decided to catch specifically this warning and ignore it (all other warnings will **not** be ignored).

leewujung replied:
I see, ok. Sounds good.

leewujung replied:
Thinking about it, this means that somewhere in the 3D array (channel x echo_range x ping_time) there is an NaN "patch" somewhere? This seems a little surprising for OOI data, although I know there are occasional, individual all-NaN pings. For the peace of mind, could you do a quick sanity check to see if there is indeed an all-NaN patch (which results into a bin that would be all NaNs)?

b-reyes replied:
@leewujung here is your sanity check that shows some pings are indeed filled with NaNs.
<img width="802" alt="Screen Shot 2022-11-23 at 12 07 27 PM" src="https://user-images.githubusercontent.com/53541061/203637352-1febe4e4-6f0a-4539-b87a-0c2cc9a37bd0.png">

leewujung replied:
Great thanks!

b-reyes commented:
@leewujung I have responded to or addressed your comments. This PR is ready for another round of review.

leewujung reviewed (approved):
@b-reyes : Everything looks good! Once you added/changed the other two thing’s in the comments, please feel free to merge this. Great work!!

b-reyes commented:
@leewujung I have addressed those comments. It looks like everything is in order here and the tests are passing. I will go ahead and merge it in shortly.
